### ISSUE TITLE
Loading in NodeJS

### ISSUE THREAD

**Issue by Adejair** [frontendbr/forum#618]
Olá galera, tudo bem com vocês? Então, comigo não está não, estou em uma situação muito complicada, vou tentar explicar o que está acontecendo comigo.

Tenho um sistema no qual eu utilizo a seguinte stack: 

* NodeJS 
* ExpressJS (session, passport.js e alguns afins) 
* MySQL
* Jade 

Eu necessito carregar alguns dados do meu banco de dados em MySQL e jogar no meu template, só que estou tendo problemas com delay, as vezes pode carregar e as vezes não, isso é questão de criar um loading, para só renderizar os dados quando estiver carregado.

Mas **EU NÃO SEI FAZER ISTO**, eu tenho uma ideia de como pode ser feito, mas não sei como fazer, vou dar ideia aqui, se vocês conseguirem me orientar =) 

Eu sei que o [plugin ](https://github.com/mysqljs/mysql) que eu usou para MySQL tem um metodo no qual eu consigo saber quando os dados foram todos carregados

```js
let query = connection.query('SELECT * from pugs');

query.on('result', (row) => {
  console.log(row);
});
```

O problema é, como eu posso mostrar um loader (template, css) enquanto os dados são carregados na tela ???

**Comment by kivervinicius** [comment]
Seu carregamento é todo em ajax? ou a tela toda somente é renderizada ao final dos dados estarem carregados?

**Comment by Adejair** [comment]
Não, não estou utilizando Ajax para carregar os dados, estou só usando os metódos do express para renderizar os dados na tela, não sei se isso é idiotice minha, é?

```js
app.get('/mypugs', (req, res) => {
	res.render('my_pug_view', {allPugs: carregarPugs()});
});
```

**Comment by suissa** [comment]
Não sei se vai te ajudar mas escrevi sobre pré-carregamento dos dados no Jade http://nomadev.com.br/mean-pre-carregando-dados-na-view-com-angularjs-jade/

ps: o artigo é de 2014 não ligue o código zuado.

**Comment by yanmagale** [comment]
A nível do JS no front end, se você mudar a sua forma de consultar a `api` , utilizando o `axios` por exemplo para fazer os requests, vc pode utilizar o evento ```onUploadProgress```

https://github.com/mzabriskie/axios/issues/629

**Comment by suissa** [comment]
A solução para o que você quer é:

- deixe seu body como visibility hidden
- requisite com AJAX seus dados
- após o recebimento de todos mude o visibility para visible

**Comment by kivervinicius** [comment]
Com ajax é tudo mais fácil, acredito que o que o @suissa falou seja o seu caso.

**Comment by suissa** [comment]
@Adejair Caso vc use só o Jade quando vc renderizar os dados já estarão na view.

Agora analisando seu código:

```js
app.get('/mypugs', (req, res) => {
	res.render('my_pug_view', {allPugs: carregarPugs()});
});
```

Creio que seja melhor você fazer assim:

```js
app.get('/mypugs', (req, res) => {
       const allPugs = carregarPugs();
	res.render('my_pug_view', { allPugs });
});
```

Caso essa função seja assíncrona é melhor vc utilizar Promise ou async/await.

**Comment by Adejair** [comment]
Obrigado pela opinião de todos no momento.

Acho que minha opção vai ser usar AJAX mesmo, não é uma boa eu renderizar os dados direto na minha view com o express. Eu tenho que passar a pensar a um longo prazo, pois a aplicação da empresa em que trabalho vai rodar em outros locais que não seja um webapp também.

Então, o que vocês me recomendam? Criar uma API Rest para os meus dados?

Se sim, é seguro chamar a requisição AJAX pela minha view? Vou precisar restringir certo tipo de dados a classes especificas, como User, Admin e afins, estes dados são muito sigilosos então preciso ter a certeza de uma segurança completa deles, o que vocês me recomendam?

@suissa @kivervinicius @yanmagale

**Comment by suissa** [comment]
Pronto isso não tem mas eu recomendo Vue ou React.

Sim é seguro chamar pela view, só não enviar dados sensíveis ehhehee.

**Comment by LFeh** [comment]
@Adejair Conforme resolver, nos avise :)

**Comment by Adejair** [comment]
Ahh sim @LFeh, eu esqueci de dar um feedback, aqui me desculpem.

Obrigado a todos que ajudaram, a forma de resolver foi a seguinte:

Como eu iria precisar chamar dados tanto na minha aplicação mobile quanto no webapp, eu decidi criar uma APIRest para poder resgastar os dados que eu preciso, de listagem e afins.

Utilizei React para renderizar os dados na tela junto com o axios para esperar o load.

A decisão da utilização do React foi pelo controle de renderização das ofertas que eu precisava, eu precisava ter dados importantes de quando os dados foram renderizados e qual pessoa estava com as ofertas aberta no momento, por isso foi importante o uso de React aqui, se alguém souber outra forma de fazer isso, compartilha aqui também :)

Obrigado a todos @LFeh @suissa @kivervinicius @yanmagale
